rds.tf

```diff
@@ -99,3 +99,180 @@ module "rds" {
+
+    principals {
+      identifiers = ["*"]
+      type        = "*"
+    }
+
+    condition {
+      test     = "StringNotEquals"
+      values   = ["aws:kms"]
+      variable = "s3:x-amz-server-side-encryption"
+    }
+  }
+
+  statement {
+    sid       = "DenyUnEncryptedObjectUploads"
+    effect    = "Deny"
+    actions   = ["s3:PutObject"]
+    resources = ["${aws_s3_bucket.rds_dumps[0].arn}/*"]
+
+    principals {
+      identifiers = ["*"]
+      type        = "*"
+    }
+
+    condition {
+      test     = "Null"
+      values   = ["true"]
+      variable = "s3:x-amz-server-side-encryption"
+    }
+  }
+
+  # Allow to read by specific IPs only if those IPs were defined
+  dynamic "statement" {
+    for_each = length(var.rds_s3_dump_allowed_ips) > 0 ? ["1"] : []
+
+    content {
+      sid = "AllowIP"
+
+      principals {
+        identifiers = ["*"]
+        type        = "*"
+      }
+
+      actions = [
+        "s3:GetObject",
+        "s3:ListBucket",
+      ]
+
+      resources = [
+        aws_s3_bucket.rds_dumps[0].arn,
+        "${aws_s3_bucket.rds_dumps[0].arn}/*",
+      ]
+
+      condition {
+        test     = "IpAddress"
+        variable = "aws:SourceIp"
+        values   = var.rds_s3_dump_allowed_ips
+      }
+    }
+  }
+}
+
+# Marry S3 bucket with policy
+resource "aws_s3_bucket_policy" "rds_dumps" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  bucket = aws_s3_bucket.rds_dumps[0].id
+  policy = data.aws_iam_policy_document.rds_dumps[0].json
+}
+
+# Policy for RDS role
+data "aws_iam_policy_document" "rds_dumps_role" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  statement {
+    sid = "AllowDumpS3BucketInteraction"
+
+    actions = [
+      "s3:PutObject",
+      "s3:GetObject",
+      "s3:ListBucket",
+      "s3:DeleteObject",
+    ]
+
+    resources = [
+      aws_s3_bucket.rds_dumps[0].arn,
+      "${aws_s3_bucket.rds_dumps[0].arn}/*",
+    ]
+  }
+}
+
+# Trust policy for RDS role
+data "aws_iam_policy_document" "rds_dumps_role_trust" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  statement {
+    sid = "TrustRDS"
+
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["rds.amazonaws.com"]
+    }
+  }
+}
+
+# Marry RDS role with policies
+resource "aws_iam_role_policy" "rds_dumps_role" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  name = "${local.rds_identifier}-rds-dumps-integration"
+  role = aws_iam_role.rds_dumps[0].id
+
+  policy = data.aws_iam_policy_document.rds_dumps_role[0].json
+}
+
+# Marry RDS with dumps role
+resource "aws_iam_role" "rds_dumps" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  name = "${local.rds_identifier}-rds-dumps-integration"
+
+  assume_role_policy = data.aws_iam_policy_document.rds_dumps_role_trust[0].json
+}
+
+resource "aws_db_instance_role_association" "this" {
+  count = local.rds_dumps_enabled ? 1 : 0
+
+  db_instance_identifier = module.rds.this_db_instance_id
+  feature_name           = "S3_INTEGRATION"
```

Comment:
Does this work with MySQL and other engines? I only used this `S3_INTEGRATION` with OracleSQL and it was done as part of an Option Group. If it works, then this is really awesome!